Invalid TypeScript types in StyleXTypes.d.ts and VarTypes.d.ts

### Describe the issue

The TypeScript types in `StyleXTypes.d.ts` and `VarTypes.d.ts` are invalid in several ways.

- Passing unknown to a symbol constraint
- Using generic types without their required arguments
- Passing less-restrictive types than the generic requires
- Passing unknown to incompatible types (strictNullChecks incompatibility)


### Expected behavior

Correct types the TypeScript compiler can validate.

### Steps to reproduce

Trying to use tsc to compile any TypeScript code that imports StyleX will fail.

```ts
import * as stylex from '@stylexjs/stylex';

export default stylex;
```


### Test case

Here is a Gist:

https://gist.github.com/AlexanderOMara/55a6e4724062e69cd21a16a5340b83f0

Attempting to build produces the following output:

```
node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:188:47 - error TS2344: Type 'unknown' does not satisfy the constraint '{ [key: string]: any; }'.

188 export type IDFromVarGroup<T extends VarGroup<unknown, unknown>> =
                                                  ~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:192:18 - error TS2314: Generic type 'CSSType' requires 1 type argument(s).

192   [key: string]: CSSType | string | { [key: string]: string };
                     ~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:216:39 - error TS2707: Generic type 'VarGroup' requires between 1 and 2 type arguments.

216 declare class ThemeKey<out VG extends VarGroup> extends String {
                                          ~~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:220:22 - error TS2344: Type 'unknown' does not satisfy the constraint '{ [key: string]: any; }'.

220   T extends VarGroup<unknown, symbol>,
                         ~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:232:26 - error TS2344: Type 'unknown' does not satisfy the constraint '{ [key: string]: any; }'.

232   TVars extends VarGroup<unknown, unknown>,
                             ~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:236:36 - error TS2344: Type 'TokensFromVarGroup<TVars>' does not satisfy the constraint '{ [key: string]: unknown; }'.

236   overrides: OverridesForTokenType<TokensFromVarGroup<TVars>>,
                                       ~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:236:55 - error TS2344: Type 'TVars' does not satisfy the constraint 'VarGroup<{}, symbol>'.
  Type 'VarGroup<unknown, unknown>' is not assignable to type 'VarGroup<{}, symbol>'.
    Type 'VarGroup<unknown, unknown>' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: {}; }>'.
      Types of property '__opaqueId' are incompatible.
        Type 'unknown' is not assignable to type 'symbol'.
          Type 'TVars' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: {}; }>'.
            Type 'VarGroup<unknown, unknown>' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: {}; }>'.
              Types of property '__opaqueId' are incompatible.
                Type 'unknown' is not assignable to type 'symbol'.

236   overrides: OverridesForTokenType<TokensFromVarGroup<TVars>>,
                                                          ~~~~~

node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:237:12 - error TS2344: Type 'TVars' does not satisfy the constraint 'VarGroup<unknown, symbol>'.
  Type 'VarGroup<unknown, unknown>' is not assignable to type 'VarGroup<unknown, symbol>'.
    Type 'VarGroup<unknown, unknown>' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: unknown; }>'.
      Types of property '__opaqueId' are incompatible.
        Type 'unknown' is not assignable to type 'symbol'.
          Type 'TVars' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: unknown; }>'.
            Type 'VarGroup<unknown, unknown>' is not assignable to type 'Readonly<{ __opaqueId: symbol; __tokens: unknown; }>'.
              Types of property '__opaqueId' are incompatible.
                Type 'unknown' is not assignable to type 'symbol'.

237 ) => Theme<TVars, ThemeID>;
               ~~~~~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:91:11 - error TS2344: Type 'T' does not satisfy the constraint 'string | 0'.
  Type 'string | number | null' is not assignable to type 'string | 0'.
    Type 'null' is not assignable to type 'string | 0'.

91   | Angle<T>
             ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:92:11 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

92   | Color<T>
             ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:93:9 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

93   | Url<T>
           ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:94:11 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

94   | Image<T>
             ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:95:13 - error TS2344: Type 'T' does not satisfy the constraint 'string | number'.
  Type 'string | number | null' is not assignable to type 'string | number'.
    Type 'null' is not assignable to type 'string | number'.

95   | Integer<T>
               ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:96:22 - error TS2344: Type 'T' does not satisfy the constraint 'string | number'.
  Type 'string | number | null' is not assignable to type 'string | number'.
    Type 'null' is not assignable to type 'string | number'.

96   | LengthPercentage<T>
                        ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:97:12 - error TS2344: Type 'T' does not satisfy the constraint 'string | number'.
  Type 'string | number | null' is not assignable to type 'string | number'.
    Type 'null' is not assignable to type 'string | number'.

97   | Length<T>
              ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:98:16 - error TS2344: Type 'T' does not satisfy the constraint 'string | number'.
  Type 'string | number | null' is not assignable to type 'string | number'.
    Type 'null' is not assignable to type 'string | number'.

98   | Percentage<T>
                  ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:99:9 - error TS2344: Type 'T' does not satisfy the constraint 'string | number'.
  Type 'string | number | null' is not assignable to type 'string | number'.
    Type 'null' is not assignable to type 'string | number'.

99   | Num<T>
           ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:100:16 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

100   | Resolution<T>
                   ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:101:10 - error TS2344: Type 'T' does not satisfy the constraint 'string | 0'.
  Type 'string | number | null' is not assignable to type 'string | 0'.
    Type 'null' is not assignable to type 'string | 0'.

101   | Time<T>
             ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:102:23 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

102   | TransformFunction<T>
                          ~

node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:103:19 - error TS2344: Type 'T' does not satisfy the constraint 'string'.
  Type 'string | number | null' is not assignable to type 'string'.
    Type 'null' is not assignable to type 'string'.

103   | TransformList<T>;
                      ~


Found 21 errors in 2 files.

Errors  Files
     8  node_modules/@stylexjs/stylex/lib/StyleXTypes.d.ts:188
    13  node_modules/@stylexjs/stylex/lib/VarTypes.d.ts:91
```

### Additional comments

When `strictNullChecks` is not enabled there are fewer errors, but still unusable.

Looks like the only way to use this library or one that depends on it is to set `"skipLibCheck": true` which is far from ideal.

--- a/packages/@stylexjs/stylex/src/types/StyleXTypes.d.ts
+++ b/packages/@stylexjs/stylex/src/types/StyleXTypes.d.ts
@@ -233,11 +233,11 @@ export type VarGroup<
 
 export type TokensFromVarGroup<T extends VarGroup<{}>> = T['__tokens'];
 
-export type IDFromVarGroup<T extends VarGroup<unknown, unknown>> =
+export type IDFromVarGroup<T extends VarGroup<{}>> =
   T['__opaqueId'];
 
 type TTokens = Readonly<{
-  [key: string]: CSSType | string | { [key: string]: string };
+  [key: string]: CSSType<null | string | number> | string | { [key: string]: string };
 }>;
 
 type UnwrapVars<T> = T extends StyleXVar<infer U> ? U : T;
@@ -254,6 +254,13 @@ type NestedVarObject<T> =
       [key: `@${string}`]: NestedVarObject<T>;
     }>;
 
+export type StyleX$DefineConsts = <
+  DefaultTokens extends TTokens,
+  ID extends symbol = symbol,
+>(
+  tokens: DefaultTokens,
+) => VarGroup<FlattenTokens<DefaultTokens>, ID>;
+
 export type StyleX$DefineVars = <
   DefaultTokens extends TTokens,
   ID extends symbol = symbol,
@@ -261,11 +268,11 @@ export type StyleX$DefineVars = <
   tokens: DefaultTokens,
 ) => VarGroup<FlattenTokens<DefaultTokens>, ID>;
 
-declare class ThemeKey<out VG extends VarGroup> extends String {
+declare class ThemeKey<out VG extends VarGroup<{}>> extends String {
   private varGroup: VG;
 }
 export type Theme<
-  T extends VarGroup<unknown, symbol>,
+  T extends VarGroup<{}>,
   Tag extends symbol = symbol,
 > = Tag &
   Readonly<{
@@ -277,7 +284,7 @@ type OverridesForTokenType<Config extends { [key: string]: unknown }> = {
 };
 
 export type StyleX$CreateTheme = <
-  TVars extends VarGroup<unknown, unknown>,
+  TVars extends VarGroup<{}>,
   ThemeID extends symbol = symbol,
 >(
   baseTokens: TVars,

--- a/packages/@stylexjs/stylex/src/types/VarTypes.js
+++ b/packages/@stylexjs/stylex/src/types/VarTypes.js
@@ -27,68 +27,66 @@ export type CSSSyntax =
   | '<transform-list>';
 
 type CSSSyntaxType = CSSSyntax;
+type InnerValue = null | string | number;
 
-interface ICSSType<+_T: string | number> {
+interface ICSSType<+_T: InnerValue> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-
-declare export class Angle<+T: string | 0> implements ICSSType<T> {
+declare export class Angle<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Color<+T: string> implements ICSSType<T> {
+declare export class Color<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Url<+T: string> implements ICSSType<T> {
+declare export class Url<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Image<+T: string> implements ICSSType<T> {
+declare export class Image<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Integer<+T: string | number> implements ICSSType<T> {
+declare export class Integer<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class LengthPercentage<+T: string | number>
-  implements ICSSType<T>
-{
+declare export class LengthPercentage<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Length<+T: string | number> implements ICSSType<T> {
+declare export class Length<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Percentage<+T: string | number> implements ICSSType<T> {
+declare export class Percentage<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Num<+T: string | number> implements ICSSType<T> {
+declare export class Num<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Resolution<+T: string> implements ICSSType<T> {
+declare export class Resolution<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class Time<+T: string | 0> implements ICSSType<T> {
+declare export class Time<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class TransformFunction<+T: string> implements ICSSType<T> {
+declare export class TransformFunction<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
-declare export class TransformList<+T: string> implements ICSSType<T> {
+declare export class TransformList<+T: InnerValue> implements ICSSType<T> {
   +value: ValueWithDefault<string>;
   +syntax: CSSSyntaxType;
 }
 
-export type CSSType<+T: null | string | number> =
+export type CSSType<+T: InnerValue> =
   | Angle<T>
   | Color<T>
   | Url<T>

